File: src/leaderboard.py
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

# TODO: tune constants
MAX_LEADERBOARD_ENTRIES = 20
UNRANKED_LABEL = "Unranked"
TOP4_PERCENT_MULTIPLIER = 100

TIER_ORDER = {
    "CHALLENGER": 9,
    "GRANDMASTER": 8,
    "MASTER": 7,
    "DIAMOND": 6,
    "EMERALD": 5,
    "PLATINUM": 4,
    "GOLD": 3,
    "SILVER": 2,
    "BRONZE": 1,
    "IRON": 0,
}

DIVISION_ORDER = {"I": 4, "II": 3, "III": 2, "IV": 1}


def formatTierLabel(rankData: Optional[Dict[str, Any]]) -> str:
    if not rankData:
        return UNRANKED_LABEL
    tier = str(rankData.get("tier", "")).title()
    division = str(rankData.get("rank", "")).upper()
    if division and division in DIVISION_ORDER:
        return f"{tier} {division}"
    return tier


def rankSortKey(rankData: Optional[Dict[str, Any]]) -> Tuple[int, int, int]:
    if not rankData:
        return (0, 0, 0)
    tier = str(rankData.get("tier", "")).upper()
    division = str(rankData.get("rank", "")).upper()
    leaguePoints = int(rankData.get("leaguePoints", 0))
    tierScore = TIER_ORDER.get(tier, 0)
    divisionScore = DIVISION_ORDER.get(division, 0)
    return (tierScore, divisionScore, leaguePoints)


def buildLeaderboardMessage(guildName: str, rows: List[Dict[str, Any]]) -> str:
    timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M UTC")
    header = f"TFT Leaderboard - {guildName}"
    lines = [header, f"Updated: {timestamp}", ""]
    lines.append("#  Player               Tier           LP   W/L     AvgPlc  Top4%")

    for index, row in enumerate(rows[:MAX_LEADERBOARD_ENTRIES], start=1):
        rankData = row.get("rank")
        tierLabel = formatTierLabel(rankData)
        leaguePoints = str(rankData.get("leaguePoints", "-")) if rankData else "-"
        wins = str(rankData.get("wins", "-")) if rankData else "-"
        losses = str(rankData.get("losses", "-")) if rankData else "-"
        avgPlacement = row.get("avgPlacement")
        top4Rate = row.get("top4Rate")
        avgPlacementLabel = f"{avgPlacement:.2f}" if isinstance(avgPlacement, float) else "-"
        top4Label = (
            f"{top4Rate * TOP4_PERCENT_MULTIPLIER:.0f}%" if isinstance(top4Rate, float) else "-"
        )
        playerName = row.get("playerName", "Unknown")
        lines.append(
            f"{index:>2} {playerName:<20.20} {tierLabel:<14.14} {leaguePoints:>4} {wins}/{losses:<5}"
            f" {avgPlacementLabel:>6} {top4Label:>6}"
        )

    content = "\n".join(lines)
    return f"```\n{content}\n```"

File: src/utils/leaderboard.js
const TIER_ORDER = [
  "CHALLENGER",
  "GRANDMASTER",
  "MASTER",
  "DIAMOND",
  "EMERALD",
  "PLATINUM",
  "GOLD",
  "SILVER",
  "BRONZE",
  "IRON",
  "UNRANKED"
];

const TIER_EMOJI = {
  CHALLENGER: "🏆",
  GRANDMASTER: "🥇",
  MASTER: "🥈",
  DIAMOND: "💎",
  EMERALD: "🟩",
  PLATINUM: "🔷",
  GOLD: "🟨",
  SILVER: "💩",
  BRONZE: "💩",
  IRON: "💩",
  UNRANKED: "❔"
};

const PLACE_EMOJI = ["🥇", "🥈", "🥉"];
const TRASH_EMOJI = "🗑️";
const DOG_FOOD_EMOJI = "🐶🥫";

function getTierEmoji(tier) {
  return TIER_EMOJI[tier] || TIER_EMOJI.UNRANKED;
}

function buildRankDisplay(entry) {
  if (!entry) {
    return `${TIER_EMOJI.UNRANKED} Unranked`;
  }
  const tier = entry.tier || "UNRANKED";
  const wins = entry.wins ?? 0;
  const losses = entry.losses ?? 0;
  const emoji = getTierEmoji(tier);
  if (tier === "MASTER" || tier === "GRANDMASTER" || tier === "CHALLENGER") {
    return `${emoji} ${tier} ${entry.leaguePoints} LP (${wins}W ${losses}L)`;
  }
  const division = entry.rank || "";
  const lp = entry.leaguePoints ?? 0;
  return `${emoji} ${tier} ${division} ${lp} LP (${wins}W ${losses}L)`.trim();
}

function rankScore(entry) {
  if (!entry) {
    return { tierIndex: TIER_ORDER.length - 1, divisionScore: 0, lp: 0 };
  }
  const tierIndex = TIER_ORDER.indexOf(entry.tier || "UNRANKED");
  const divisionMap = { I: 4, II: 3, III: 2, IV: 1 };
  const divisionScore = divisionMap[entry.rank] || 0;
  return { tierIndex, divisionScore, lp: entry.leaguePoints ?? 0 };
}

function sortLeaderboard(results) {
  results.sort((a, b) => {
    const scoreA = rankScore(a.rankEntry);
    const scoreB = rankScore(b.rankEntry);

    if (scoreA.tierIndex !== scoreB.tierIndex) {
      return scoreA.tierIndex - scoreB.tierIndex;
    }
    if (scoreA.divisionScore !== scoreB.divisionScore) {
      return scoreB.divisionScore - scoreA.divisionScore;
    }
    return scoreB.lp - scoreA.lp;
  });
  return results;
}

function formatLeaderboard(results, options = {}) {
  if (!results.length) {
    return "No registered players yet.";
  }
  const header = options.header || "**TFT Leaderboard**";
  const lastIndex = results.length - 1;
  const lines = results.map((result, index) => {
    const rankText = result.error ? `Error: ${result.error}` : buildRankDisplay(result.rankEntry);
    const riotLabel = `${result.riotId} (${result.region})`;
    const isLastPlace = index === lastIndex && results.length > 1;
    const place = isLastPlace ? DOG_FOOD_EMOJI : PLACE_EMOJI[index] || TRASH_EMOJI;
    const name = result.riotId.split("#")[0];
    return `${place} ${name} - ${riotLabel} - ${rankText}`;
  });
  return `${header}\n${lines.join("\n")}`;
}

module.exports = {
  buildRankDisplay,
  formatLeaderboard,
  sortLeaderboard
};
